### PR TITLE
Start working on OpenApi definition and docs

### PR DESCRIPTION
**Pull request by Lothiraldan**

### CODE REVIEW

**Comment by Lothiraldan** [comment]
One thing I'm having trouble documenting is when the bug itself is not available as we are currently returning a 200 status code. It might be better to return a 40x status code instead.

Also the answer schema for the batch API is very hard to define because we might have successfully predicted bug in the answer, in-processing bug and unavailable bugs in the same answer payload.

I see three possibilities:
- I can try to find a way to define the answer payload to handle all the cases
- We can move the retrieval of the result to the non-batch API, so we would batch X bugs and for each bug the consumer will ask the "normal" endpoint about them.
- We only document the answer through examples.

**Comment by marco-c** [comment]
> We can move the retrieval of the result to the non-batch API, so we would batch X bugs and for each bug the consumer will ask the "normal" endpoint about them.

This might be too slow if you have a lot of bugs to retrieve.

> We only document the answer through examples.

This might be the easiest way. Maybe a short explanation of how it works followed by examples.

File: http_service/app.py [hunk 1]
```diff
@@ -8,16 +8,42 @@
...
 
-from .models import classify_bug
+from bugbug import get_bugbug_version
+
+from .models import MODELS_NAMES, classify_bug
 
 API_TOKEN = "X-Api-Key"
 
+API_DESCRIPTION = """
+This is the documentation for the BubBug http service, the platform for Bugzilla Machine Learning project.
```
Comment (marco-c) [inline http_service/app.py]:
```suggestion
This is the documentation for the BubBug http service, the platform for Bugzilla Machine Learning projects.
```

File: http_service/app.py [hunk 2]
```diff
@@ -8,16 +8,42 @@
...
+
+# Introduction
+
+This service can be used to classify a given bug using one of pre-trained model.
```
Comment (marco-c) [inline http_service/app.py]:
```suggestion
This service can be used to classify a given bug using a pre-trained model.
```

File: http_service/app.py [hunk 3]
```diff
@@ -8,16 +8,42 @@
...
+You can classify a single bug or a batch of bugs.
+The classification happens on background so you need to call back the service for getting the results.
```
Comment (marco-c) [inline http_service/app.py]:
```suggestion
The classification happens in the background so you need to call back the service for getting the results.
```

File: http_service/app.py [hunk 4]
```diff
@@ -85,15 +142,52 @@ def get_bug_classification(model_name, bug_id):
...
 
 
-@application.route("/<model_name>/predict/<bug_id>")
+@application.route("/<model_name>/predict/<int:bug_id>")
+@cross_origin()
 def model_prediction(model_name, bug_id):
+    """
+    ---
+    get:
+      description: Classify a single bug using given model, answer either 200 if the bug is processed or 202 if at least the bug is being processed
```
Comment (marco-c) [inline http_service/app.py]:
```suggestion
      description: Classify a single bug using given model, answer either 200 if the bug is processed or 202 if the bug is being processed
```

File: http_service/app.py [hunk 5]
```diff
@@ -85,15 +142,52 @@ def get_bug_classification(model_name, bug_id):
...
+      summary: Classify a single bug id
```
Comment (marco-c) [inline http_service/app.py]:
```suggestion
      summary: Classify a single bug
```

File: http_service/app.py [hunk 6]
```diff
@@ -85,15 +142,52 @@ def get_bug_classification(model_name, bug_id):
...
+          type: integer
+          example: 123456
+      responses:
+        200:
+          description: A single bug prediction
+          content:
+            application/json:
+              schema: BugPrediction
+        202:
+          description: A temporary answer for bug being processed
```
Comment (marco-c) [inline http_service/app.py]:
```suggestion
          description: A temporary answer for the bug being processed
```

File: http_service/app.py [hunk 7]
```diff
@@ -110,13 +204,124 @@ def model_prediction(model_name, bug_id):
...
+        ```<br/><br/>
+
+        And you will get the following output once the bugs are available:
+        <br/>
+        ```
+        {"bugs": {"456": {"extra_data": {}, "index": 0, "prob": [0], "suggestion": ""}}}
+        ```<br/><br/>
+
+        Please be aware that the results might be interleaved, so the
+        following output is valid:
```
Comment (marco-c) [inline http_service/app.py]: What did you mean here?

File: http_service/app.py [hunk 8]
```diff
@@ -110,13 +204,124 @@ def model_prediction(model_name, bug_id):
...
+        <br/>
+        ```
+        {"bugs": {"123": {"available": False}, "456": {"extra_data": {}, "index": 0, "prob": [0], "suggestion": ""}}}
+        ```
+      summary: Classify a batch of bugs id
```
Comment (marco-c) [inline http_service/app.py]:
```suggestion
      summary: Classify a batch of bugs
```

File: http_service/app.py [hunk 9]
```diff
@@ -110,13 +204,124 @@ def model_prediction(model_name, bug_id):
...
+                      index: 0
+                      prob: [0]
+                      suggestion: string
+                    789012:
+                      extra_data: {}
+                      index: 0
+                      prob: [0]
+                      suggestion: string
+        202:
+          description: A temporary answer for bug being processed
```
Comment (marco-c) [inline http_service/app.py]:
```suggestion
          description: A temporary answer for bugs being processed
```

File: http_service/models.py [hunk 10]
```diff
@@ -109,7 +109,8 @@ def classify_bug(
 
     model = get_model(model_name)
 
-    model_extra_data = model.get_extra_data()
+    # model_extra_data = model.get_extra_data()
+    model_extra_data = {}
```
Comment (marco-c) [inline http_service/models.py]: Why this change?

File: http_service/templates/doc.html [hunk 11]
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>BugBug doc</title>
+    <!-- needed for adaptive design -->
```
Comment (marco-c) [inline http_service/templates/doc.html]: Nit: remove this comment

File: http_service/templates/doc.html [hunk 12]
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>BugBug doc</title>
```
Comment (marco-c) [inline http_service/templates/doc.html]:
```suggestion
    <title>BugBug Documentation</title>
```

File: http_service/app.py (hunk 7 above)
Comment (Lothiraldan) [reply http_service/app.py]: I meant that for a batch, some of the bugs might be available, some of the bugs might still be processing and some of the bugs results might be ready

Comment (marco-c) [reply http_service/app.py]: Oh, OK, can you specify that?
Something like:
```
Please be aware that the results might be interleaved, so the following output, where a bug is returned and another one is still being processed, is valid:
```

**Comment by Lothiraldan** [comment]
@marco-c Did you take a look at the generated documentation ? It's on http://localhost:5000/doc when running locally

File: http_service/models.py (hunk 10 above)
Comment (Lothiraldan) [reply http_service/models.py]: Unrelated change, checking that everything is working as it should without it

File: http_service/app.py (hunk 7 above)
Comment (Lothiraldan) [reply http_service/app.py]: Ok done

File: http_service/models.py (hunk 10 above)
Comment (Lothiraldan) [reply http_service/models.py]: Everything is good
